0023-communication-protocol.adoc

```diff
@@ -0,0 +1,388 @@
+
+More generally speaking:
+Given knowledge of the alpha and beta ledger and the role, a message can be unambiguously identified as the redeem or refund identity of an actor depending on whether the message is incoming or outgoing.
+
+This allows us to slim down the protocols to the bare minimum of sending over an identity.
+
+=== Finalizing a swap
+
+One of the downsides of splitting the communication for a single swap up into several protocols is the fact that we don't have a natural "end" of the communication.
+If all of the communication were to take place within a single protocol, closing the substream could act as "I am done communicating".
+
+Following a multiple protocol approach, the answer to how we communicate this concern is another protocol.
+
+Introducing: `/comit/swap/finalize/1.0.0`
+
+[source,json]
+----
+{
+  "swap_id": "713b0c93-53a1-46b2-9418-de8d46750b79"
+}
+----
+
+Both parties are expected to use this protocol after they have sent all the data they consider necessary for starting the execution of the swap.
+Once each party sent a finalize message AND received one, implementations should consider the swap finalized and start with the execution phase (i.e. watch for transactions, emit actions, etc)
+
+== Discussion points
+
+This section includes discussion topics that remain to be answered.
+
+=== Should we use multihash for the swap digest
+
+<<Announce-message,Section 3.2>> introduces the concept of a swap digest.
+In order to be forwards compatible with different hash algorithms, it proposes to encode the merkle root as a multihash.
+That way, the hash is self-describing and the receiving party knows, which hash algorithm needs to be used to compute the digest.
+
+Do we need the flexibility of changing the hash function that is used to compute the swap digest without a breaking change to the protocol?
+We could always just create `/comit/swap/announce/2.0.0` that defines a different hash function.
+
+=== How to compute the swap digest?
+
+It is still to be defined, how exactly the swap digest will be computed.
+In particular, we need to define:
+
+* the order in which the tree of information about the swap is arranged
+* which parts of the swap are included in the merkle tree
+
+Especially the last point is important to discuss.
+Consider the following:
+
+Given the above protocols, it is should be possible to realize an implementation of COMIT that only requires a single user interaction (on both sides) before the swap starts.
+Considering our current implementation, we would still generate transient bitcoin identities to use in the HTLC and only pay to the user's wallet upon redeeming.
+Both of these requirements combined raises the question:
+how does cnd know, which of the parameters submitted via the API are "negotiated" (i.e. also known to the other party) and which ones where added locally by the user?
+Only the "negotiated" ones must be added to the swap digest, otherwise the digests won't match.
```

Comment:
I like the proposed `announce` protocol for its simplicity.
Two comments, just for the sake of awareness: 

1. We have to very carefully select which parameters are part of the digest to avoid **collisions** (i.e. same ledgers+assets+amounts+protocol, but different identities); for a high-frequency trader it is likely to have a lot of swaps in progress. Note that this issue is a general issue that is independent of calculation of the digest, but has to be taken into account when actually designing the digest.

I assume that a (unlikely) collision (i.e. two swaps with the same digest being in the pre-announce state at the same time) is in the responsibility of the application on top of COMIT?

2. There is no way to "give feedback" to the other party any more. When sending the swap parameters we see what the other party sends and can provide feedback on what is missing/wrong. When sending a computed digest we only find a match or we fail. There is no feedback whatsoever. We can see this as advantage (more resilient to spam) and disadvantage (no feedback to a different COMIT implementation).

Reply:
Regarding 1)

Yes, the swap digest is not an identifier. Hence, any application using this communication protocol has to come with its own means of uniquely identifying data that flows through the system. For cnd, this likely means assigning a local ID to any request made on the REST API.

Any party receiving such an announcement message with the same digest as a previous one MUST treat them as different swaps and assign a new ID.

Regarding 2)

I would suggest to add messages for feedback later on, once we have clear needs to avoid a similar mess that we currently have with the unused decline reason.

Reply:
I added a section about collision resistance.